odxtools/statemachine.py

```diff
@@ -0,0 +1,174 @@
+        fsm = StateMachine(ecu, ecu.state_charts.my_state_chart)
+
+        for raw_request in (executor := fsm.execute(ecu.services.my_service,
+                                                    param1="hello",
+                                                    param2=123)):
+            # send raw request to the ECU (usually using ISO-TP or DoIP), and
+            # receive the response
+            await send_to_ecu(raw_request)
+            while odxtools.is_response_pending(raw_response := await receive_from_ecu()):
+                pass
+
+            executor.send(raw_response)
+
+        print(f"Request send: {fsm.succeeded}")
+        ```
+
+        """
+        if service.request is None:
+            odxraise("Services without requests are not allowed in this context")
+            self.succeeded = False
+            return
+
+        # the service can be executed if any of the specified
+        # precondition states is fulfilled. (TODO: correct?)
+        if service.pre_condition_state_refs is not None:
+            if not any(
+                    x.applies(self, service.request.parameters, service_params)
+                    for x in service.pre_condition_state_refs):
+                # if all preconditions which are applicable are
+                # invalid (i.e., they evaluate to False), we must not
+                # execute the service.
+                self.succeeded = False
+                return
+
+        raw_req = service.request.encode(**service_params)
+        # ask the calling code to send the request to the ECU and
+        # report back the reply
+        raw_resp = yield raw_req
+        if raw_resp is None:
+            raise RuntimeError("The calling code must send back a reply")
+        elif isinstance(raw_resp, (bytes, bytearray)):
+            for decoded_resp_msg in self.diag_layer.decode_response(raw_resp, raw_req):
+                for stransref in service.state_transition_refs:
+                    # we only execute the first applicable state
+                    # transition: The spec seems to imply a
+                    # deterministic state machine and chaining
+                    # transistions most likely is not what the user
+                    # expects. (The spec seems to be a bit loose on
+                    # this front...)
+                    if stransref.execute(self, decoded_resp_msg.coding_object.parameters,
+                                         decoded_resp_msg.param_dict):
+                        self.succeeded = True
+                        yield b''
+                        return
+        else:
+            resp_object = service.positive_responses[0]
+            for stransref in service.state_transition_refs:
+                if stransref.execute(self, resp_object.parameters, raw_resp):
+                    self.succeeded = True
+                    yield b''
+                    return
+
+        self.succeeded = True
+        yield b''
```

Comment:
Why yield empty bytes?

Reply:
I was surprised about this as well, but returning outright causes a `StopIteration` to be raised here. instead of terminating the loop after a single iteration. For whatever reason, the second yield does not get passed to the calling code, see the [unit test](https://github.com/mercedes-benz/odxtools/pull/398/files#diff-032f4e3b4207885cc9f9a556d079a94d23c497c353da952e672357c10b12de65R59-R66).